src/network_dynamic.c

```diff
@@ -0,0 +1,654 @@
+#include <string.h>
+#include <ctype.h>
+#include "os_utils.h"
+#include "os_pic.h"
+#include "network.h"
+#include "shared_context.h"
+#include "common_utils.h"
+#include "apdu_constants.h"
+#include "write.h"
+#include "hash_bytes.h"
+#include "lcx_hash.h"
+#include "lcx_sha256.h"
+#include "public_keys.h"
+#ifdef HAVE_LEDGER_PKI
+#include "os_pki.h"
+#endif
+
+#define P1_NETWORK_CONFIG 0x00
+#define P1_NETWORK_ICON   0x01
+#define P1_GET_INFO       0x02
+#define P2_FIRST_CHUNK    0x01
+#define P2_NEXT_CHUNK     0x00
+
+#define MAX_ICON_LEN 1024
+
+#define TYPE_DYNAMIC_NETWORK   0x08
+#define NETWORK_STRUCT_VERSION 0x01
+
+#define BLOCKCHAIN_FAMILY_ETHEREUM 0x01
+
+// Tags are defined here:
+// https://ledgerhq.atlassian.net/wiki/spaces/FW/pages/5039292480/Dynamic+Networks
+enum {
+    TAG_STRUCTURE_TYPE = 0x01,
+    TAG_STRUCTURE_VERSION = 0x02,
+    TAG_BLOCKCHAIN_FAMILY = 0x51,
+    TAG_CHAIN_ID = 0x23,
+    TAG_NETWORK_NAME = 0x52,
+    TAG_TICKER = 0x24,
+    TAG_NETWORK_ICON_HASH = 0x53,
+    TAG_DER_SIGNATURE = 0x15,
+};
+
+// Signature context structure
+typedef struct {
+    uint8_t sig_size;
+    const uint8_t *sig;
+    cx_sha256_t hash_ctx;
+} s_sig_ctx;
+
+#ifdef HAVE_NBGL
+typedef struct {
+    uint16_t received_size;
+    uint16_t expected_size;
+    uint16_t slot;
+} network_payload_t;
+
+// Global variable to store the network icons
+typedef struct {
+    uint8_t bitmap[MAX_ICON_LEN];
+    uint8_t hash[CX_SHA256_SIZE];
```

Comment:
Since we already received the hash along with the network configuration, couldn't we do something like this to match the two ? `icon_hash_from_network_config == sha256(icon_bitmap)`
It would save on RAM at the (small) cost of a little runtime computation.

Reply:
This is how it works :wink: